rfcs/rfc-bld-20230302-1-3p-development.md

```diff
@@ -0,0 +1,215 @@
+*   Github packages - The storage service for packages. Artifacts created and uploaded to this service in OCI container format
+*   AWS Elastic Container Service (ECS) - A container service using auto scaled hosts for self-hosted runners  
+*   AWS Elastic Container Registery (ECR) - A service that hosts pre-built container images  
+*   S3 - A storage device that holds packages for internal repos
+*   AWS Cloudfront - A content delivery network (CDN) which caches packages for internal repos at global sites
+
+$~$
+
+Additional Considerations
+-------------------------
+The advantages to this approach:
+
+*   All scripts can be contained in the .github/workflow folder, which will follow along with any fork or branch.
+*   Github Actions is invoked within the scope of the repo, fork, or branch, which decentrializes its output
+*   Github Actions environments are containerized, so that a developer can reproduce builds locally
+*   Github packages in OCI format come with "free" vulnerability scanning, supply chain protection (containers can be signed), and traceability (the artifact hash can be correlated to other public packages)
+*   For public repos, this infrastructure is "free" of cost, with some nebulous limit but seems higher than the sum of our current packages. The only time costs come into play is for private repos, which already come with a bucket of Actions minutes and Packages storage space.
+
+$~$
+
+There's some cons however:
+*   Because the development package output is decentralized, we won't have a central "development" S3 bucket. This means a contributor will have to ensure that they add the Github packages URL for their fork to the the list of CDN urls to be pulled via cmake.
+*   Github Actions uses a containerized infrastructure, with limited specs - something on the order of [2 CPUs with 7 GB RAM](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) with 19~30GB of SSD storage. Large packages like QT and Python may take hours to build. There may be ways to utilize Linux Foundation's hardware as "agents" (called self-hosted runners) for Github Actions if we need higher specs
+*   Support for versions of OS is on a ["window" of releases](https://github.com/actions/virtual-environments#available-environments). Some OS versions will fall out of support, which will have to be augmented by self-hosted runners
+*   We're at the mercy of Github for any support and if Github itself goes down, then users pulling 3p from an installed version of O3DE may get blocked. A supplementary CDN hosting on [o3debinaries.org](http://o3debinaries.org) could be a potential backup. This is controlled by the CDN list in [https://github.com/o3de/o3de/blob/development/cmake/3rdPartyPackages.cmake#L32](https://github.com/o3de/o3de/blob/development/cmake/3rdPartyPackages.cmake#L32)

```

Comment:
If we're concerned about the availability of the GitHub's package distribution blocking installs, we can still use our Cloudfront/S3 solultion for hosting 3p packages while automating the deployment with GitHub actions.

Reply:
Ok great, I was going to suggest we perform a phased rollout during the migration and looks like that is already the case. 

Since we will already pay for the S3 storage, will migrating the package hosting to github's package distribution provide us with other benefits over the current solution? I'm wondering if we can save us the effort of creating a new driver if we continuing using our current CDN solution.

Reply:
Yes, there's benefits and some nuances I should explain here: 
1. We don't have a public "development" bucket or CDN in the O3DE AWS account. To replicate the current infrastructure for public contribution would significantly increase cost to O3DF, as well as increasing complexity of managing access to the bucket. Github Packages is free for public projects (and contributor forks) and can be controlled by our current Github permissions.
2. Using Github Packages is visible and reproducible to the public. The public currently has no view of the packaging or uploaded artifacts in the "production" S3 bucket until after it is pulled during the engine/project build process
3. Github Packages outbound data transfer and storage costs would be free. Using the current S3/Cloudfront CDN as a infrequently used backup would reduce O3DF's AWS costs (we would potentially reduce the storage tier as well).